### PR TITLE
Code coverage improvement for buffer wrappers

### PR DESCRIPTION
**Pull request by dparanic**

### CODE REVIEW

File: test/general/buffer_wrapper.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#if TEST_DPCPP_BACKEND_PRESENT
+
+    std::size_t size = 1000;
+    sycl::buffer<uint32_t> buf{size};
+    test_buffer_wrapper test{};
+
+    test(oneapi::dpl::begin(buf), oneapi::dpl::end(buf), size);
+    test(oneapi::dpl::begin(buf, sycl::write_only), oneapi::dpl::end(buf, sycl::write_only), size);
+    test(oneapi::dpl::begin(buf, sycl::write_only, sycl::noinit), 
+            oneapi::dpl::end(buf, sycl::write_only, sycl::noinit), size);
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:70]: Could you please align identation with the previous line?

File: test/general/buffer_wrapper.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include "support/utils.h"
+#include <vector>
+
+
+//This macro is required for the tests to work correctly in CI with tbb-backend.
+#if TEST_DPCPP_BACKEND_PRESENT
+struct test_buffer_wrapper
+{
+    template <typename Iterator1, typename Iterator2>
+    void operator()(const Iterator1& begin1, const Iterator2& end1, std::size_t size)
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:17]: Iterators usually pass by copy

File: test/general/buffer_wrapper.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        std::fill(oneapi::dpl::execution::dpcpp_default,begin1,end1,1);
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:19]: please, add spaces after commas

File: test/general/buffer_wrapper.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        auto begin_host1 = TestUtils::get_host_pointer(begin1);
+        auto end_host1 = TestUtils::get_host_pointer(end1);
+
+        for(auto it = begin_host1; it!= end_host1; it=it+1)
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:23]: please, add spaces before and after `!=`, `=`, `+`

File: test/general/buffer_wrapper.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            EXPECT_TRUE(*it == 1,"wrong effect of reduce with sycl_iterator");
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:25]: reduce --> fill?

File: test/general/buffer_wrapper.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        it = end1;
+
+        EXPECT_TRUE(it == end1, "wrong effect of iterator's copy assignment operator");
+        EXPECT_TRUE(it - size == begin1, "wrong effect of iterator's operator - integer");
+        EXPECT_TRUE(begin1 + size == end1, "wrong effect of iterator's operator + integer");
+        EXPECT_TRUE(end1 - begin1 == size, "wrong effect of iterator's operator - iterator");
+
+        auto buf = begin1.get_buffer();
+        auto begin_host2 = TestUtils::get_host_pointer(oneapi::dpl::begin(buf));
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:46]: I'd suggest to get accessors of buffer using SYCL API, not our internal.

File: test/general/buffer_wrapper.pass.cpp (hunk 1 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Done

File: test/general/buffer_wrapper.pass.cpp (hunk 2 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Aligned it with common practice.

File: test/general/buffer_wrapper.pass.cpp (hunk 3 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Done

File: test/general/buffer_wrapper.pass.cpp (hunk 4 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Done

File: test/general/buffer_wrapper.pass.cpp (hunk 5 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Fixed

File: test/general/buffer_wrapper.pass.cpp (hunk 6 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Applied it.

File: test/general/buffer_wrapper.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        it = end1;
+
+        EXPECT_TRUE(it == end1, "wrong effect of iterator's copy assignment operator");
+        EXPECT_TRUE(it - size == begin1, "wrong effect of iterator's operator - integer");
+        EXPECT_TRUE(begin1 + size == end1, "wrong effect of iterator's operator + integer");
+        EXPECT_TRUE(end1 - begin1 == size, "wrong effect of iterator's operator - iterator");
+
+        std::fill(oneapi::dpl::execution::dpcpp_default, begin1, end1, 1);
+        auto buf = begin1.get_buffer();
+        EXPECT_TRUE(buf.get_count() == size, "wrog effect of iterator's method get_buffer");
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:38]: Here we check that size of the buffer getting from the `get_buffer()` method equals `size`. Let's think what is a better way to do the `get_buffer()` check.
Maybe it's better to get access and see that pointers of original buffer and the buffer from the `get_buffer()` method on underlined data are the same?

File: test/general/buffer_wrapper.pass.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        auto begin_host = sycl::host_accessor<typename Iterator::value_type, 1, sycl::access_mode::read>{buf}.get_pointer();
+        auto end_host = sycl::host_accessor<typename Iterator::value_type, 1, sycl::access_mode::read>{buf}.get_pointer() + size;
+
+        for(auto it = begin_host; it != end_host; it = it + 1)
+        {
+            EXPECT_TRUE(*it == 1, "wrong effect of fill algorithm with sycl_iterator");
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:45]: Could you please clarify why we need to check `fill` algorithm? We have a separate test for `fill`

File: test/general/buffer_wrapper.pass.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+#if TEST_DPCPP_BACKEND_PRESENT
+
+    std::size_t size = 1000;
+    sycl::buffer<uint32_t> buf{size};
+    test_buffer_wrapper test{};
+
+    test(oneapi::dpl::begin(buf), oneapi::dpl::end(buf), size);
+    test(oneapi::dpl::begin(buf, sycl::write_only), oneapi::dpl::end(buf, sycl::write_only), size);
+    test(oneapi::dpl::begin(buf, sycl::write_only, sycl::noinit), oneapi::dpl::end(buf, sycl::write_only, sycl::noinit), size);
+    test(oneapi::dpl::begin(buf, sycl::noinit), oneapi::dpl::end(buf, sycl::noinit), size);
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:63]: SYCL 2020 doesn't have `sycl::noinit`. It has `sycl::no_init`. Could you, please, fix it in the test and possible in our implementation if it's needed?

File: test/general/buffer_wrapper.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        EXPECT_TRUE(it - size == begin1, "wrong effect of iterator's operator - integer");
+        EXPECT_TRUE(begin1 + size == end1, "wrong effect of iterator's operator + integer");
+        EXPECT_TRUE(end1 - begin1 == size, "wrong effect of iterator's operator - iterator");
+
+        std::fill(oneapi::dpl::execution::dpcpp_default, begin1, end1, 1);
+        auto buf = begin1.get_buffer();
+        EXPECT_TRUE(buf.get_count() == size, "wrog effect of iterator's method get_buffer");
+
+        auto begin_host = sycl::host_accessor<typename Iterator::value_type, 1, sycl::access_mode::read>{buf}.get_pointer();
+        auto end_host = sycl::host_accessor<typename Iterator::value_type, 1, sycl::access_mode::read>{buf}.get_pointer() + size;
```
[the remark this answers is not in the document]
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Applied

File: test/general/buffer_wrapper.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        EXPECT_TRUE(it == end1, "wrong effect of iterator's copy assignment operator");
+        EXPECT_TRUE(it - size == begin1, "wrong effect of iterator's operator - integer");
+        EXPECT_TRUE(begin1 + size == end1, "wrong effect of iterator's operator + integer");
+        EXPECT_TRUE(end1 - begin1 == size, "wrong effect of iterator's operator - iterator");
+
+        std::fill(oneapi::dpl::execution::dpcpp_default, begin1, end1, 1);
+        auto buf = begin1.get_buffer();
+        EXPECT_TRUE(buf.get_count() == size, "wrog effect of iterator's method get_buffer");
+
+        auto begin_host = sycl::host_accessor<typename Iterator::value_type, 1, sycl::access_mode::read>{buf}.get_pointer();
```
[the remark this answers is not in the document]
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Simplified

File: test/general/buffer_wrapper.pass.cpp (hunk 7 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Yes, I think it is better way to check method. Applied.

File: test/general/buffer_wrapper.pass.cpp (hunk 9 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp:54]: I believe, it is better to fix it in separated pull request.

File: test/general/buffer_wrapper.pass.cpp (hunk 8 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Erased this check.

**Review by capatober — APPROVED** [review]

**Pull request merged**